### PR TITLE
Feature/types

### PR DESCRIPTION
**Pull request by brianignacio5**
This PR re writes the ESPTOOL-JS in TypeScript and preparation for NPM packaging and publishing.

Proposed features here:
- Re write existing code in TypeScript and add types for targets and methods used.
- Re write example `index.js` for applied changes.
- Add Github CI for publishing in NPM.
- Add scripts for lint the existing typescript code.
- Add rollup script to bundle the resulting JS (after npm run build) into a single `bundle.js` used in the example for browser purposes. 
- Removed XTerm within ESPLoader code and opted for dependency injection of Terminal reference. This allows the existing code to use different terminals systems in JavaScript/Typescript.
- Added package.json references and compilation from TypeScript to JavaScript so EsptoolJS could be used as a Node module with type references. 

Things left out:

- Testing (How should we provide mocking for the Transport class to test ESPLoader functionality ?
- NPM account configuration to publish.

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  publish:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          node-version: '14.x'
+          registry-url: 'https://registry.npmjs.org'
+      - name: Install dependencies and build 🔧
+        run: npm install && npm run build
```
Comment (balloob) [inline .github/workflows/publish.yml:16]: Let's use `npm ci` so it will strictly adhere to the lockfile.

```suggestion
        run: npm ci && npm run build
```

File: rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  input: 'lib/index.js',
+  output: {
+    name: "esptooljs",
+    file: 'bundle.js',
+    format: 'es',
+  },
+  plugins: [
+    resolve(),
+    commonjs(),
+    babel({ exclude: 'node_modules/**', babelHelpers: "runtime", skipPreflightCheck: true }),
```
Comment (balloob) [inline rollup.config.js:19]: What is Babel used for? TypeScript can already convert to a target JavaScript version

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "ES5",
```
Comment (balloob) [inline tsconfig.json:4]: All browser support modern JavaScript. We should leave it up to the users of this lib to decide what level of JavaScript they want to consume.

```suggestion
    "module": "es2020",
    "target": "es2019",
```

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    "declaration": true,
+    "allowSyntheticDefaultImports": true,
+    "outDir": "./lib",
+    "moduleResolution": "node",
+    "lib": ["ES2020", "DOM"]
```
Comment (balloob) [inline tsconfig.json:9]: Let's import the helpers. It requires that you add `tslib` to the dependencies but it allows for further JavaScript de-duplication.

```suggestion
    "lib": ["ES2020", "DOM"],
    "importHelpers": true
```

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "target": "ES5",
+    "declaration": true,
+    "allowSyntheticDefaultImports": true,
```
Comment (balloob) [inline tsconfig.json:6]: What dependency uses this ?

File: src/targets/esp8266.ts [hunk 6]
```diff
@@ -179,83 +185,97 @@ export default class ESP8266ROM {
...
+      this._d2h(mac[4]) +
+      ":" +
+      this._d2h(mac[5])
+    );
+  };
 
-  static get_erase_size = function(offset, size) {
-      return size;
-  }
+  public get_erase_size = function (offset: number, size: number) {
```
Comment (balloob) [inline src/targets/esp8266.ts:278]: If you are going to change all these functions, let's set them correctly for classes:

```suggestion
  public get_erase_size(offset: number, size: number) {
```

File: src/espLoader.ts [hunk 7]
```diff
@@ -0,0 +1,1008 @@
...
+  chip: ROM;
+  IS_STUB: boolean;
+  FLASH_WRITE_SIZE: number;
+
+  constructor(
+    public transport: Transport,
+    private baudrate: number,
+    private terminalCleanFn: () => void,
+    private terminalWriteLineFn: (data: string) => void,
+    private terminalWriteFn: (data: string) => void,
```
Comment (balloob) [inline src/espLoader.ts:81]: I think that the old way that allowed passing in a terminal object was nicer. Easier to reason about. Why did you break it up?

File: src/espLoader.ts [hunk 8]
```diff
@@ -0,0 +1,1008 @@
...
+    private rom_baudrate = 115200,
+  ) {
+    this.IS_STUB = false;
+    this.chip = new ESP32ROM();
```
Comment (balloob) [inline src/espLoader.ts:85]: This is incorrect? You haven't actually spoke to the device here.

File: src/espLoader.ts [hunk 9]
```diff
@@ -0,0 +1,1008 @@
...
+      throw new ESPError("Soft resetting is currently only supported on ESP8266");
+    } else {
+      // running user code from stub loader requires some hacks
+      // in the stub loader
+      this.command(this.ESP_RUN_USER_CODE, undefined, undefined, false);
+    }
+  }
+}
+
+export default ESPLoader;
```
Comment (balloob) [inline src/espLoader.ts:1008]: Let's try to avoid default exports. Named exports are easier to manage and easier to track as it's always same name.

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+import ESPLoader from "./espLoader";
```
Comment (balloob) [inline src/index.ts:1]: let's lowercase the `L` in `espLoader` filename.

File: src/targets/rom.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    return size;
+  }
+
+  FLASH_SIZES: { [key: string]: number } = {
+    "1MB": 0x00,
+    "2MB": 0x10,
+    "4MB": 0x20,
+    "8MB": 0x30,
+    "16MB": 0x40,
+  };
```
Comment (balloob) [inline src/targets/rom.ts:52]: This is a duplicate of above?

```suggestion
  FLASH_SIZES = FLASH_SIZES;
```

File: rollup.config.js (hunk 2 above)
Comment (brianignacio5) [reply rollup.config.js:19]: Babel is used to modify ES6 JS code into [browser friendly code](https://babeljs.io). Without this, JS code was not working properly in the browser.

File: src/espLoader.ts (hunk 7 above)
Comment (brianignacio5) [reply src/espLoader.ts:81]: Because we want to use other terminals beside xTerm. For example, use this Esptool-js in Eclipse Theia and vscode extension. The previous implementation was using terminal functions within espLoader, I think injecting dependency gives more flexibility. Passing the entire Terminal object is also no go, because methods like `writeLn` are not the same for all terminal frameworks.

Comment (balloob) [reply src/espLoader.ts:81]: Right, but it could be an interface that consumers can wrap their terminal in.

```ts
new ESPLoader({
  clean() {
    myTerm.doClean();
  }

  write(data) {
    myTerm.doWrite(data)
  }
})
```

File: rollup.config.js (hunk 2 above)
Comment (balloob) [reply rollup.config.js:19]: Ok 👍

In that case we can have TypeScript output modern JS as suggested in https://github.com/espressif/esptool-js/pull/53#discussion_r1028122793

File: index.js [hunk 12]
```diff
@@ -1,27 +1,27 @@
...
+const tableBody = document.getElementById('tableBody');
 const table = document.getElementById('fileTable');
 const alertDiv = document.getElementById('alertDiv');
 
-//import { Transport } from './cp210x-webusb.js'
-import { Transport } from './webserial.js'
-import { ESPLoader } from './ESPLoader.js'
-import { ESPError } from './error.js'
+// import { Transport } from './cp210x-webusb.js'
+import * as esptooljs from "./bundle.js";
```
Comment (balloob) [inline index.js:20]: I hope in the future we can just import this from unpkg like we suggest for ESP Web Tools:

https://esphome.github.io/esp-web-tools/#add-website

**Comment by balloob** [comment]
You can delete `webserial.js`

File: package.json [hunk 13]
```diff
@@ -1,8 +1,52 @@
...
+  ],
+  "scripts": {
+    "build": "npm run clean && tsc",
+    "clean": "rimraf lib bundle.js",
+    "format": "prettier --write \"src/**/*.ts\" \"src/**/*.js\"",
+    "lint": "eslint . --ext .ts",
+    "lintAndFix": "eslint . --ext .ts --fix",
+    "prepare": "npm run build",
+    "test": "echo \"Error: no test specified\"",
+    "prepublishOnly": "npm run test && npm run lint",
```
Comment (balloob) [inline package.json:19]: FYI lint currently fails, but I suggest to run prettier in follow-up PR which also adds a CI step so every PR is checked.

File: src/esploader.ts [hunk 14]
```diff
@@ -0,0 +1,1009 @@
...
+import ESP8266ROM from "./targets/esp8266";
+
+const MAGIC_TO_CHIP: { [key: number]: ROM } = {
+  0x00f01d83: new ESP32ROM(),
+  0x6921506f: new ESP32C3ROM(), // ECO 1+2
+  0x1b31506f: new ESP32C3ROM(), //ECO 3
+  0x09: new ESP32S3ROM(),
+  0x000007c6: new ESP32S2ROM(),
+  0xfff0c101: new ESP8266ROM(),
+};
```
Comment (balloob) [inline src/esploader.ts:18]: Why did you change these from dynamic imports to always imported? This means that the main bundle is way large. The web can easily dynamically import things.

If for some reason you do not want this in bundle.js (it used to work there previously), you can use the Rollup option [`inlineDynamicOption`](https://rollupjs.org/guide/en/#outputinlinedynamicimports)

**Review by balloob — APPROVED** [review]

File: index.js (hunk 12 above)
Comment (brianignacio5) [reply index.js:20]: I think we can release the bundle to unpkg too !

File: src/esploader.ts (hunk 14 above)
Comment (brianignacio5) [reply src/esploader.ts]: I think the size would be the same because the bundle will still include all targets with `inlineDynamicImports`. The alternative would mean to include those targets js files to deployment, which will against the idea of having a bundle single js file. We could probably optimize this later to bundle for specific targets ?

Comment (balloob) [reply src/esploader.ts]: The NPM package should include 2 builds. A build that is just the transpiled version. Leave it up to the consumer of this lib to decide how it should deal with bundling of this lib, including de duplication of transpiration artifacts. 

The second build can be a single bundle.js build, which can use inlineDynamicImports to generate a single file. However, I think that’s overkill and will actually hurt performance. Users want to connect fast, and it’s a waste to let their computers download ALL chips (including all future ESP chips!) before they can connect one. 

When do you see the bundle.js being used directly? Because I thought that the goal was to publish it on NPM and then it’s up to the consumer to deal with it?

**Review by balloob — APPROVED** [review]

File: src/esploader.ts (hunk 14 above)
Comment (brianignacio5) [reply src/esploader.ts]: Yes @balloob this is what we have at the moment. NPM package includes the `lib` folder with transpiled JS files with TS types and the `bundle.js` made from those JS files which I think is useful for browser deployment since it is better to have single minified file (yet to minify it xD) like releasing it in unpkg and other CDNs.

File: .github/workflows/publish.yml [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+      - uses: actions/setup-node@v3
+        with:
+          node-version: '14.x'
+          registry-url: 'https://registry.npmjs.org'
+      - name: Install dependencies and build 🔧
+        run: npm ci && npm run build
+      - name: Run lint 
+        run: npm run lint
+      - name: Run tests 
+        run: npm run test
```
Comment (igrr) [inline .github/workflows/publish.yml:20]: @brianignacio5 should we run build & lint & test steps also on pull requests? (And package etc only for tags)

File: src/esploader.ts (hunk 14 above)
Comment (balloob) [reply src/esploader.ts]: Unpkg works with modules and relative imports just fine. That’s what we do for ESP Web Tools and it looks like this https://unpkg.com/browse/esp-web-tools@9.0.4/dist/web/ (this split is the result of the dynamic imports in esptool.js)

File: .github/workflows/publish.yml [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+      - name: Package module
+        run: npm pack
+      - name: Determine version
+        id: version
+        run: echo "version=${GITHUB_REF:11}" >> $GITHUB_OUTPUT
```
Comment (igrr) [inline .github/workflows/publish.yml:25]: I've noticed that we use different logic to determine the package version here and in ci.yml.

Do we need to check that the tag name matches package.json version name, or does `npm publish` take care of that? (What happens if we tag a new version but forget to bump the version number in package.json?)

Comment (brianignacio5) [reply .github/workflows/publish.yml:25]: The difference is that in any regular PR the version is not part of the tag here. We are getting the version from package.json because `npm pack` will generate a `esptool-js-<version>.tgz` which we use to create a PR artifact that could be used for review/testing. 

In the publish.yml workflow, we use the GITHUB_REF because the workflow is triggered on tags with `v*` to do the release the module in npm. While npm publish doesn't bump the version in package.json, but if the tag already exist in the GitHub repository this workflow would not run (i.e if version already exists) so new version is required for release.

We could modify it so this step gets the version from package.json, but the triggering mechanism remains as described before.

Comment (igrr) [reply .github/workflows/publish.yml:25]: The only case I'm worried about is if we tag a new version but forget to update the version number in package.json.

To me it seems like an additional check in the workflow for tags would solve this. That is, extract the version from package.json and verify if it's the same as the tag name.
